src/Psalm/Type/Atomic/TString.php

```diff
@@ -15,4 +26,12 @@ public function getKey()
     {
         return 'string';
     }
+
+    /**
+     * @return string
+     */
+    public function getId()
+    {
+        return $this->values ? 'string(\'' . implode('\',\'', array_keys($this->values)) . '\')' : 'string';
```

Comment:
Long-term, this would need to be quoted (So would array shape keys)

A whitelist might include `\0x7f-\0xff` and anything that isn't a special character. 

E.g. `$x = "O')|string('" would be rendered as having a type of `string('O')|string('')

Reply:
Yes, this definitely needs to be toughened up a little